lib/Doctrine/DBAL/Platforms/SQLServer2012Platform.php

```diff
@@ -103,17 +103,29 @@ protected function doModifyLimitQuery($query, $limit, $offset = null)
         // Queries using OFFSET... FETCH MUST have an ORDER BY clause
         // Find the position of the last instance of ORDER BY and ensure it is not within a parenthetical statement
         // but can be in a newline
-        $matches      = [];
-        $matchesCount = preg_match_all('/[\\s]+order\\s+by\\s/im', $query, $matches, PREG_OFFSET_CAPTURE);
-        $orderByPos   = false;
+        $matches          = [];
+        $matchesCount     = preg_match_all('/[\\s]+order\\s+by\\s/im', $query, $matches, PREG_OFFSET_CAPTURE);
+        $shouldAddOrderBy = true;
         if ($matchesCount > 0) {
-            $orderByPos = $matches[0][$matchesCount - 1][1];
+            // ORDER BY instance may be in a subquery after ORDER BY
+            // e.g. SELECT col1 FROM test ORDER BY (SELECT col2 from test ORDER BY col2)
+            // if in the searched query ORDER BY clause was found where
+            //  number of open parentheses after the occurrence of the clause is equal to
+            //  number of closed brackets after the occurrence of the clause,
+            // it means that ORDER BY is included in the query being checked
+            $orderByAlreadyExists = false;
+            $tempMatchesCount     = $matchesCount;
+            while ($tempMatchesCount && ! $orderByAlreadyExists) {
+                $orderByPos           = $matches[0][--$tempMatchesCount][1];
+                $openBracketsCount    = substr_count($query, '(', $orderByPos);
+                $closedBracketsCount  = substr_count($query, ')', $orderByPos);
+                $orderByAlreadyExists = ($openBracketsCount - $closedBracketsCount === 0);
+            }
+
+            $shouldAddOrderBy = ! $orderByAlreadyExists;
```

Comment:
Would it make sense to move all this logic into a separate `private function shouldAddOrderBy(): bool`? It could help avoid awkward variable names like `$tmpMatchesCount` and reduce the complexity of this method.